### ISSUE TITLE
Mutual TLS: permit empty organization unit in client cert

### ISSUE THREAD

**Issue by vito** [cloudfoundry-incubator/credhub#25]
As mentioned in [the mutual TLS docs](https://github.com/cloudfoundry-incubator/credhub/blob/master/docs/mutual-tls.md#limitations), client certificates must contain an organization unit field in the form of `app:some-guid`.

The docs already make it sound like this'll be fixed at some point, I just wanted to +1 that from a user's perspective. :) This bit us when testing Concourse's CredHub integration, and we ended up having to generate our own certs in code rather than just use BOSH's variables feature. (See https://github.com/concourse/topgun/commit/b5300a15384c8a2c5226b9c35ba55cc4c911c9f7.) I would like to just not have to fill in anything for that field.

**Comment by crawsible** [comment]
Hey @vito! Sorry for the delay.

We use the OU with that pattern to establish the identity of a user authenticating via mTLS. That means we won't be able to allow connections from clients presenting certificates that don't include that OU pattern, though we might support certs without OUs if we're able to identify the user in some other way. Are you planning to connect to CredHub over mTLS and then authenticate via OAuth2 or something along those lines?

**Comment by vito** [comment]
@crawsible Concourse uses CredHub without really having a notion of a 'user', though I suppose that could change - it's really just how our integration is implemented at the moment. I just liked to have mTLS as an alternative authentication scheme so that we could avoid deploying and configuring a UAA + client id/secret in our test suite, though I could see users wanting to have a similar setup.

**Comment by athornton2012** [comment]
@vito is this issue still relevant for the concourse team?

**Comment by vito** [comment]
@athornton2012 I believe so yes.

**Comment by johnpfield** [comment]
@crawsible Just wondering why OU for identity rather than SubjectName? or SubjectAltName.  Usually the subject of the certificate is the holder of the private key.

**Comment by athornton2012** [comment]
@johnpfield I suspect that we use the OU for identity because the diego instance identity certificates use the OU to hold the app guid. 

@crawsible perhaps it makes sense that if ACLs are disable in CredHub, then we don't care about the OU.

@vito when Concourse supports multiple teams using the same deployment with CredHub, I suspect ACLs will be enabled. At that point would it make more sense to introduce another concept of a user (i.e. mtls-app, uaa-user, uaa-client, concourse-team) ? Or perhaps make this configurable in CredHub?

**Comment by crawsible** [comment]
@johnpfield: @athornton2012's right. We built mTLS auth specifically for the Secure Service Credentials use-case. Service brokers want to grant access to creds in CredHub at the app-level, not the instance level, and Diego puts the app identifier in the OU and instance identifier in the SN. We want to be able to treat instances of apps as a single identity, and Diego wants to provision identity credentials as granularly as possible, and that's how we arrived here.

I'm somewhat of the opinion that disabled ACL's days are numbered and we should work towards eliminating the ability to disable ACLs, so it feels a little weird for me to spend cycles on disabled ACLs mode.

Ultimately, I think the authn certs need to include _some_ form of identifier, but there's no reason it needs to be in the OU. I'd be open to expanding the x509 fields we accept as a source of identity to SN, for example, though we'd likely continue to impose some sort of structure on the content of the field, so that we can be confident when its value is intentionally representing a CredHub identity.
